src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +673,482 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+        _as.request_abort();
+        co_await _waiter_gate.close();
+
+        co_return std::move(result);
+    }
+
+private:
+    shard_local_fetch_context _ctx;
+    ss::gate _waiter_gate;
+    ss::abort_source _as;
+
+    // This signals when an index has been added to `_request_indexes` or when
+    // `_as` aborts.
+    ss::condition_variable _has_completed_waiters;
+    // This contains indexes into `_ctx.requests` and is populated by raft
+    // consensus waiters when an offset increases for a partition.
+    std::vector<size_t> _request_indexes;
+    // Contains the last observed value of `consensus->last_visible_index()` for
+    // every request in `_ctx.requests`. Its used to register waiters with
+    // `consensus->visible_offset_monitor()`.
+    std::vector<model::offset> _last_visible_indexes;
+
+    struct query_results {
+        std::vector<model::offset> last_visible_indexes;
+        // Indicates if any `read_result` in `results` has an error.
+        bool has_error;
+        std::vector<read_result> results;
+        size_t total_size;
+    };
+
+    ss::future<query_results>
+    query_requests(std::vector<ntp_fetch_config> requests) {
+        // The last visible indexes need to be populated before partitions
+        // are read. If they are populated afterwards then the
+        // last_visible_index could be updated after the partition is read,
+        // but before the index is queried. This leads to a race condition
+        // where we're waiting for the last_visible_index to change even
+        // though the partition has data we haven't read.
+        //
+        // Note that `last_visible_index` is used instead of the
+        // `commit_offset` as the commit offset won't change when partitions
+        // are produced to without acks=all. The `last_visible_index` more
+        // closely corresponds to the Kafka high watermark as well.
+        std::vector<model::offset> last_visible_indexes{requests.size()};
+        std::vector<std::tuple<size_t, model::partition_id>> errored_partitions;
+        size_t total_size{0};
+        bool has_error{false};
+
+        for (size_t i = 0; i < requests.size(); i++) {
+            const auto& req = requests[i];
+            auto part = _ctx.mgr.get(req.ktp());
+            if (!part) {
+                errored_partitions.emplace_back(i, req.ktp().get_partition());
+                continue;
+            }
+            auto consensus = part->raft();
+            if (!consensus) {
+                errored_partitions.emplace_back(i, req.ktp().get_partition());
+                continue;
+            }
+            last_visible_indexes[i] = consensus->last_visible_index();
+        }
+
+        // A read_result needs to be returned for every partition. Hence,
+        // the function can't return before calling
+        // `fetch_ntps_in_parallel`.
+        std::vector<read_result> results = co_await fetch_ntps_in_parallel(
```

Comment:
It's possible for us to have detected an error for a partition above, but then succesfully read from it here in fetch_ntps_in_parallel (I assume, hard to see how that could be avoided).

Will that cause a problem? We will not have populated the last_visible_offset for one, and query_results.has_error may be true while none of query_results.results indicate any error. Maybe that's fine? Personally I'd prefer that once we may the decision to error out a partition, we stick with it and don't try to fetch from it (e.g., by flagging or removing the request prior to passing it down to fetch_ntps_in_parallel).

Reply:
It won't cause a problem. If for a partition `fetch_ntps_in_parallel` succeeds, but `last_visible_offset` fails then the function overrides the successful result with the failure and returns from the fetch. Both need to succeed for the partition to be successfully fetched from. I'd expect the case where `last_visible_offset` fails but `fetch_ntps_in_parallel` succeeds to happen very rarely.

Reply:
Well just looking briefly down the function it doesn't override the whole result, it just sets the error code and partition, but the result object will still have all the other state from the fetch, right? Just below that we add it's size to `total_size` without checking the error code, so total_size includes size for errored results, is that intentional and OK? Looking at the result object functions like `has_data()` will return true. So you would need to carefully audit the entire lifetime of this object to ensure every caller is handling the case of this "franken object" in a suitable way.

Basically what I'm getting at is that it's better to have only 2 states for the result object "really errored" and "successful with data", and not have to think about this third case of "successful but we overwrote the error code".

Reply:
At least in all the cases where I've overridden a successful result with an error-ed one I've replaced the entire `read_result` not just changed the error code. I.e, in the lines below the one you commented I do;
```  
        for (auto [i, partition] : errored_partitions) {                                                                                                                     
              results[i] = read_result(error_code::not_leader_for_partition);                                                                                                              
              results[i].partition = partition;                                                                                                                                               
          }                                                                                                                                                                                   
```
Which replaces the entire read result with an empty one that has the `not_leader_for_partition` error code.

Reply:
Let me double check that `total_size` will be correct everywhere though.

Reply:
So there was a slight issue here https://github.com/redpanda-data/redpanda/pull/15328/files/b5cfb3ffb14acdcc2834d7afe53493ea73f981de#diff-ccfd5e131023aa80d73e854c51f6dd29973c5fe5003c004aa4487f637319879dR927-R937

I need to subtract the current size of the read from the partition from total_size before I change it to an empty result. I.e, 
```cpp                  
 total_size -= results[r_i].data_size_bytes();                                                                                                                               
```

Reply:
> Which replaces the entire read result with an empty one

Ah, I'm not sure how I missed this, I thought we were only assigning two fields but you are right the whole object is replaced. Looks good and sorry for the noise.